### ISSUE TITLE
release binaries should be statically linked

### ISSUE THREAD

**Issue by jane0009** [lun-4/awtfdb#60]
### nixos moment
on NixOS, awtfdb does not work out of the box because library paths are hard-coded.

currently I can use autoPatchelfHook to fix this just fine, but the hope is that eventually i can build awtfdb locally and entirely circumvent this issue.

to facilitate this: it would be great to have some way to either
- specify the library paths manually, or
- automatically search for the libraries if they don't exist at the default path, akin to pkg-config or similar

i have no idea how feasible the second option is, so the easiest would probably just be the first - i.e. check if some file specifying library paths exists, and then use those over the default paths.

**Comment by lun-4** [comment]
there are two problems here. the first problem is "awtfdb binaries do not work by default on nixos", which is fixable by making awtfdb bulid inside nix (like you said, `autoPatchelfHook` for the binary builds, and probably something nixpkgs already provides to build any kind of zig-ware).

this issue happens because when i build in my Void system, it hardcodes a path to the dynamic loader, as it should. that path is then read by the kernel, and in nixos' case, the loader specified will resolve the `.so` symbols to `/nix` paths, setup relevant memory pages, etc, etc, and then run your binary code. however, if you're not pointing to nix's loader, errors happen. `autoPatchelfHook` changes the loader path so it works.

example:
```
$ file zig-out/bin/ainclude
zig-out/bin/ainclude: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, with debug_info, not stripped
```

`/lib64/ld-linux-x86-64.so.2` is checked by the kernel *before any of our code can run* (i may be mistaken on this, but it's on the ELF headers, which are static). to try to attach custom auto-loader logic to find a loader beforehand (and maybe even rewrite the loader path) is well beyond my abilities (this is getting into [cosmopolitan libc](https://github.com/jart/cosmopolitan) territory). attempt to build binaries that are more distro-independent is a good idea, though (in the future, this should all be linked statically, which removes the need for a dynamic loader entirely, making the binary only need the kernel. this inflates the size of the builds and binaries, which leads us to the second problem -- why are awtfdb release binaries not statically linked if it's more pain?).

the second problem is "the regex tag inferrer inside `ainclude` hardcodes /lib", which is not a fatal issue, you only trip yourself on it if you run `ainclude --infer-tags regex --regex-use-exif ...` as that attempts to load GraphicsMagick .so files from `/usr/lib` using `dlopen()`, other than that the entire system works. maybe there could be an `AWTFDB_LIB_PATH` environment variable that you hardcode on your nix system, don't know a good UX for that. this all still requires the binaries to be dynamically loaded because GraphicsMagick depends on your libc (which is a dynamic library). this would be fixed by statically linked builds, as mentioned in the first problem.

tl;dr static builds are the future.

**Comment by jane0009** [comment]
static builds, understandable! i think when i originally wrote this ticket, especially in the context of the second option, i was thinking about it as "check where the libraries are located at build time", because i think build.zig still has a hardcoded check for places in /usr/lib - not sure how resolvable that in particular is. i recently learned about nix's magical utility called substituteInPlace for fixing this exact issue at build time so i might mess around with that at some point, but that's not a strictly awtfdb related task

edit: just looked, immediately noticing a hardcoded reference to /usr/include but that's it. not sure what other issues i was having before but those may be related to zigmod
